cmd/peer-rest-client.go

```diff
@@ -423,51 +423,49 @@ func (client *peerRESTClient) SignalService(sig serviceSignal) error {
 }
 
 // Trace - send http trace request to peer nodes
-func (client *peerRESTClient) Trace(doneCh chan struct{}, trcAll bool) (chan []byte, error) {
-	ch := make(chan []byte)
+func (client *peerRESTClient) Trace(traceCh chan interface{}, doneCh chan struct{}, trcAll bool) {
 	go func() {
-		cleanupFn := func(cancel context.CancelFunc, ch chan []byte, respBody io.ReadCloser) {
-			close(ch)
-			if cancel != nil {
-				cancel()
-			}
-			http.DrainBody(respBody)
-		}
 		for {
 			values := make(url.Values)
 			values.Set(peerRESTTraceAll, strconv.FormatBool(trcAll))
+
 			// get cancellation context to properly unsubscribe peers
 			ctx, cancel := context.WithCancel(context.Background())
 			respBody, err := client.callWithContext(ctx, peerRESTMethodTrace, values, nil, -1)
 			if err != nil {
-				//retry
+				// Retry the failed request.
 				time.Sleep(5 * time.Second)
-				select {
-				case <-doneCh:
-					cleanupFn(cancel, ch, respBody)
-					return
-				default:
+			} else {
+				dec := gob.NewDecoder(respBody)
+
+				go func() {
+					<-doneCh
+					// cancel() the request so that bio.Scan() can return.
+					cancel()
+				}()
+
+				for {
+					var info trace.Info
+					if err = dec.Decode(&info); err != nil {
+						break
+					}
+					select {
+					case traceCh <- info:
+					default:
```

Comment:
yes. The convention here and at pubsub's publisher is "channel sender does not care about slow receivers" because if it needs to care we will have to implement it this way on the sender's side:

```
go func() {
   select {
    case ch <- entry:
    case <-doneCh:
        return
}()
```
in a goroutine. Downside is - misbehaving/slow subscribers (or because of network issues) will cause goroutine build up on the senders side. On a high load production setup this might start impacting memory.

So chan receivers need to use buffered chan to make sure that they don't miss the non-blocking sends.

I tested with high load, none of the trace events were missed due to the current arrangement.

Reply:
Better not log because slow receivers might cause flood of logs (and tracing is not critical to the functioning of minio anyhow)